### ISSUE TITLE
ERROR: ReadBlockFromDisk: OpenBlockFile failed for CBlockDiskPos(nFile=-1, nPos=0)

### ISSUE THREAD

**Issue by soeren-b-c** [gandrewstone/bitcoin#386]
I have a node that has stopped twice in the last couple of hours.

The last message in debug.log is:

ERROR: ReadBlockFromDisk: OpenBlockFile failed for CBlockDiskPos(nFile=-1, nPos=0)

**Comment by nomnombtc** [comment]
Same here... That assert code in main.cpp:5706 also exists in the bitcoin-core 0.12 branch, hm.

```
2017-03-21T19:35:19.826069329Z ERROR: ReadBlockFromDisk: OpenBlockFile failed for CBlockDiskPos(nFile=-1, nPos=0)
2017-03-21T19:35:19.841467037Z bitcoind: main.cpp:5706: bool ProcessMessage(CNode*, std::string, CDataStream&, int64_t): Assertion `!"cannot load block from disk"' failed.
```

**Comment by mmouse-** [comment]
Same here, I could offer a coredump.
edit: Full node, debian jessie, kernel 3.16, 64bit

**Comment by buuser** [comment]
Same here, pruning node. Are others pruning or full?

**Comment by nomnombtc** [comment]
Fullnode. On my other (full)node, which has not crashed so far, I see things like this in the logfile at the moment, maybe related?:

```
2017-03-21 20:56:40 Misbehaving: 192.160.102.164:9569 (0 -> 100) BAN THRESHOLD EXCEEDED
2017-03-21 20:56:40 ProcessMessages(get_xthin, 46 bytes) FAILED peer=23115
2017-03-21 21:02:45 Misbehaving: 188.219.232.164:53362 (0 -> 100) BAN THRESHOLD EXCEEDED
2017-03-21 21:02:45 ProcessMessages(get_xthin, 46 bytes) FAILED peer=23126
2017-03-21 21:04:39 Misbehaving: 197.231.221.211:24075 (0 -> 100) BAN THRESHOLD EXCEEDED
2017-03-21 21:04:39 ProcessMessages(get_xthin, 46 bytes) FAILED peer=23130
2017-03-21 21:07:29 Misbehaving: 185.29.8.211:37492 (0 -> 100) BAN THRESHOLD EXCEEDED
2017-03-21 21:07:29 ProcessMessages(get_xthin, 46 bytes) FAILED peer=23136
```

**Comment by greatwolf** [comment]
Running 64-bit BU 1.0.1.1 on Win7, pruned to 30gb. No crash yet.

**Comment by ftrader** [comment]
@nomnombtc : Those are the typical msgs from last pre-1.0.1.1 exploit (after patching) - those attackers sending invalid xthin got banned

**Comment by KlausT** [comment]
Same error under Windows (bitcoin-qt 64bit):
```
Assertion failed!
bitcoin-qt.exe
File: main.cpp, Line 5706

Expression: !"cannot load block from disk"
```
Last lines of debug.log:
```
2017-03-21 20:31:12 Acceptable block: ver:20000000 time:1490128262 size: 998120 Tx:1994 Sig:5337
2017-03-21 20:31:12 BLOCK_DOWNLOAD_WINDOW is 8 MAX_BLOCKS_IN_TRANSIT_PER_PEER is 1
2017-03-21 20:31:12 UpdateTip: new best=00000000000000000115096b766fa7e59b110aa73a998094e932a9781bfad676  height=458305  log2_work=86.160782  tx=206777763  date=2017-03-21 20:31:02 progress=1.000000  cache=72.4MiB(57205tx)
2017-03-21 20:31:12 UpdateTip: 27 of last 100 blocks have unexpected version. One example: 0x20000002
2017-03-21 20:32:23 Misbehaving: 91.223.82.156:33916 (0 -> 100) BAN THRESHOLD EXCEEDED
2017-03-21 20:32:23 ProcessMessages(get_xthin, 46 bytes) FAILED peer=6985
2017-03-21 20:36:55 socket send error 'Eine bestehende Verbindung wurde softwaregesteuert durch den Hostcomputer abgebrochen.  (10053)' to 198.251.83.19:36060 (6994)
2017-03-21 20:38:46 ERROR: ReadBlockFromDisk: OpenBlockFile failed for CBlockDiskPos(nFile=-1, nPos=0)

```
By the way: 3 nodes are banned now.

**Comment by afk11** [comment]
How long til they acknowledge this one..

**Comment by h0jeZvgoxFepBQ2C** [comment]
Looks like core fixed it already in 0.13 :)

**Comment by nomnombtc** [comment]
@ftrader Yeah I know it looks like from the earlier attack, just found it strange my log was pretty quiet then suddenly, around the time my other node crashed, it was full with this stuff.

**Comment by satoshisSockpuppet** [comment]
No idea if I'm right but it looks like that just attacks pruned nodes.

line 5705.. in main.cpp seems to use assert as an error for missing blocks on disk.
line 4852 was introduced with f9ec3f0f apparently to avoid exactly that situation.

if (send && (mi->second->nStatus & BLOCK_HAVE_DATA))

**Comment by ShadowOfHarbringer** [comment]
> No idea if I'm right but it looks like that just attacks pruned nodes.

I don't think so. My node was never pruned and crashed as well.

**Comment by ftrader** [comment]
We have reports of full nodes (unpruned) crashing, but @satoshisSockpuppet comment seems to go in the right direction w.r.t. the asserts ...

**Comment by ddyzhang** [comment]
For what it's worth, _all_ of my Windows x64 full nodes crashed (2 out of 2), but none of my Ubuntu 16.04 nodes have been affected (0 out of 2).

**Comment by KlausT** [comment]
My second node on a different Windows machine is still running

**Comment by ShadowOfHarbringer** [comment]
I am running my node on x64 Gentoo and it crashed. It is also causing (after restart) high system load, but that may be coincidence.

EDIT:
I confirm that my I/O load is definately higher than normal. The client is doing more hard-drive activity than usual (checked on iotop).

**Comment by zander** [comment]
guys, you can restart your node with the following option which likely fixes it for now;

`-use-thinblocks=0`

**Comment by ShadowOfHarbringer** [comment]
> -use-thinblocks=0

Ok ,I am gonna try the workaround.

**Comment by greatwolf** [comment]
Does this issue affect Classic? I know Classic added XThin awhile back mostly using BU's code.

**Comment by ddyzhang** [comment]
Higher-than-normal IO not yet confirmed on my end. Seems to be normal according to iotop.

**Comment by ShadowOfHarbringer** [comment]
> Higher-than-normal IO not yet confirmed on my end. Seems to be normal according to iotop.

Very well, I am going to verify this now that I started node with workaround. The higher load maybe also because of other factors.

**Comment by h0jeZvgoxFepBQ2C** [comment]
Maybe you should just update to 0.13? should be fixed there?

**Comment by nomnombtc** [comment]
@lichtamberg depends what causes it, the assert itself is still present in bitcoin master, just in a different file: https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1040

**Comment by satoshisSockpuppet** [comment]
@nomnombtc 
No, there is a check before that, that is missing in the second 
assert(!"cannot load block from disk")
in line 5706 in main.cpp

**Comment by ShadowOfHarbringer** [comment]
After watching IOTOP for another few minutes while trying starting client with and without workaround I think it was most probably a coincidence.

**Comment by kristovatlas** [comment]
Looks to me like this assert was introduced in the very first xthin commit for main.cpp:

source:
https://github.com/BitcoinUnlimited/BitcoinUnlimited/blob/2c9e2c62f5bf4369f4bff7ba783ae32432de4f68/src/main.cpp#L5231

commit:
https://github.com/BitcoinUnlimited/BitcoinUnlimited/commit/2c9e2c62f5bf4369f4bff7ba783ae32432de4f68

(Note to other reviewers looking to compare to Core: Core renamed main.cpp to validation.cpp in https://github.com/bitcoin/bitcoin/commit/76faa3cdfedbd3fc91be4ecfff77fc6dc18134fb)

**Comment by is55555** [comment]
If you upgrade to Core this has been fixed for ages.

**Comment by kristovatlas** [comment]
@nomnombtc: That assert is there, but this check appears to prevent it from being executed unless the db is corrupted rather than simply missing the block from the db.

https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L1035
```cpp
// Pruned nodes may have deleted the block, so check whether
                // it's available before trying to send.
                if (send && (mi->second->nStatus & BLOCK_HAVE_DATA))
                {
                    // Send block from disk
                    CBlock block;
                    if (!ReadBlockFromDisk(block, (*mi).second, consensusParams))
                        assert(!"cannot load block from disk");
```

Edit: This was also true for Core 0.12:
https://github.com/bitcoin/bitcoin/blob/0.12/src/main.cpp#L4450

**Comment by ftrader** [comment]
@is55555 But I don't want to downgrade my possible block size at the same time as fixing this bug.

**Comment by h0jeZvgoxFepBQ2C** [comment]
@ftrader why? there are no big blocks at the moment?

**Comment by ShadowOfHarbringer** [comment]
@lichtamberg

>  there are no big blocks at the moment?

And there never will be. That is the problem with Core.

**Comment by is55555** [comment]
There will never be? So you assume Bitcoin is always going to be Core. I have to agree with that prediction.

**Comment by ShadowOfHarbringer** [comment]
> So you assume Bitcoin is always going to be Core. I have to agree with that prediction.

No, your logic is flawed. By following what I have said it is clear I was talking about Core. Note that I never used the word "Bitcoin". 
And that's because Core is not Bitcoin anymore. At least not for me.

**Comment by h0jeZvgoxFepBQ2C** [comment]
At least there will never be big blocks, if the nodes keep crashing all the time...

**Comment by LivInTheLookingGlass** [comment]
Can we please keep conversation restricted to the issue at hand? There are lots of places to talk Bitcoin politics, but perhaps the issue tracker isn't one of them.

**Comment by is55555** [comment]
Actually I was wrong above when I said this was fixed for ages in Core.

Core never had this bug. It's in the xthin code (like the remote zero-day).

**Comment by kristovatlas** [comment]
Hi, please take your low-effort shit-posting to Reddit.

**Comment by ffallah** [comment]
My debian node crashed as well AGAIN. I am not running a node anymore until BU team get their sh** together. this is unacceptable.

**Comment by lastbattle** [comment]
Yup, mine crashed too :( let me check out the logs

**Comment by kFTY** [comment]
My Raspberry Pi node crashed too.
System: Ubuntu 16.04.2 LTS (GNU/Linux 4.4.38-v7+ armv7l)

Debug.log before it stops:

```
2017-03-22 01:31:13 Acceptable block: ver:20000000 time:1490146211 size: 998153 Tx:2364 Sig:4305
2017-03-22 01:31:13 BLOCK_DOWNLOAD_WINDOW is 8 MAX_BLOCKS_IN_TRANSIT_PER_PEER is 1
2017-03-22 01:31:13 UpdateTip: new best=0000000000000000009438446ed0a3d0a437a38a395d9893ba4b4dfa34b625e0  height=458334  log2_work=86.16177  tx=206843152  dat$
2017-03-22 01:31:13 UpdateTip: 28 of last 100 blocks have unexpected version. One example: 0x20000002
2017-03-22 01:33:55 ERROR: ReadBlockFromDisk: OpenBlockFile failed for CBlockDiskPos(nFile=-1, nPos=0)
```

After the crash, bitcoin-cli cannot get response, however, the database is still locked. I restart the raspberry pi and it works again.

**Comment by is55555** [comment]
Since this feature is completely broken, you can try disabling it and that should stop this particular crash.

I believe you can do this:

-use-thinblocks=0

Good luck

EDIT: I mean, this is better than running unsigned effing binaries for such a software...

**Comment by ShadowOfHarbringer** [comment]
> EDIT: I mean, this is better than running unsigned effing binaries for such a software...

Seriously, please keep your political discussion out of this forum. There is enough of this stuff on reddit.

**Comment by is55555** [comment]
Do you think that was political? It was a pretty solid recommendation and if you don't think so I seriously doubt your judgement.

Running an unsigned binary for software like this is bonkers. Much better to disable thinblocks like I just recommended.

**Comment by levino** [comment]
If you are using the Ubuntu ppa to install the binaries, you are always just running compiled binaries while trusting the ppa admin that the code is the result of the compile. I honestly see no difference in the two scenarios: Both times you trust the admin of the ppa.

**Comment by is55555** [comment]
Why on Earth would you use Ubuntu ppa? Don't assume I'm mentally insane just because I'm posting here right now.

**Comment by sickpig** [comment]
@gandrewstone @AndrewClifford  what about closing this issue?

**Comment by sickpig** [comment]
Closed by release 1.0.1.2
